src/v/cluster/log_eviction_stm.cc

```diff
@@ -60,15 +60,6 @@ ss::future<> log_eviction_stm::write_raft_snapshots_in_background() {
     /// snapshots as close to effective_start_offset as possible.
     auto gh = _gate.hold();
     while (!_as.abort_requested()) {
-        try {
```

Comment:
I actually liked the old structure of the code because it meant that we do no work if there's nothing to do, instead of now calling `do_write_raft_snapshot` unconditionally. Not sure if this no-ops, but at least just inspecting the code previously, the behavior was obvious

Consider something like (though really rough, and needs an update to consider the max collectible offset)

```
optional<time> recheck_timeout = nullopt;
while (!aborted) {
  if (recheck_timeout.has_value()) {
    reap_condition.wait(recheck_timeout.value());
  } else {
    reap_condition.wait();
  }
  evict_until = ...
  index_lb = ...
  if (index_lb && _raft->last_snapshot_index() < index_lb) {
    do_write_raft_snapshot(index_lb);
  }
  if (... the last snapshot successfully truncated all past the storage/delete-records truncation points...) {
    recheck_timeout = nullopt;
  } else {
    recheck_timeout = 5s;
  }
}
```

Reply:
yeah it will no-op, theres two cases:
1. index_lb is null because theres no place to truncate, so `do_write_snapshot()` is avoided
2. Within `do_write_snapshot()` we do nothing if index_lb < raft snapshot

Reply:
Gotcha. It'd be nice to not have to read through to see what no-ops (and rather have no-ops surface in the business logic of this loop), but not a blocking opinion at all

Reply:
yeah i agree, going to change this too